Source: https://github.com/ADVANTECH-Corp/linux-imx.git
Context: imx8LBVF0170 (branch adv_6.6.23_2.0.0)

commit af78e96d161d6d807935924c6df912aa52e7e7e3 (tag: imx8LBVF0106)
Author: kuihong.xin <kuihong.xin@advantech.com.cn>
Date:   Fri Jun 6 17:59:18 2025 +0800
 
    [imx8mm][RSB-3730][DISPLAY] Modify for dsi-lvds panel bridge & add gpio-pwm
 
commit 605789a569e288b22adf4c0a71821010c297670f
Author: Hank.Hsueh <Hank.Hsueh@advantech.com.tw>
Date:   Tue May 27 10:24:35 2025 +0800
 
        [imx8mp][ALL][drivers] Update device support display enhancements
        - Add support for auo_g070vw01v0, auo_g150xgel05, auo_g215hvn01, innolux_r190ece, lg_lp097x02, lt9211_800x480, lt9211_1024x768, lt9211_1280x800, lt9211_1920x1080 panels
        - Add GPIO handling for gpio_bklt_enable_desc
 
commit e8fcbda7552ca3ea516bc88de02c33a22d934a44
Author: Hank.Hsueh <Hank.Hsueh@advantech.com.tw>
Date:   Tue Dec 3 14:23:07 2024 +0800
 
    [ALL][DISPLAY] driver: Add the LVDS power on sequence
    commit 9e1727c7de8aa4e025dfbe300d75c1a1a77b0729

20260313: update for lf-6.18.2-1.0.0

Observation: ad hoc patch

Upstream-Status: Inappropriate [other]
Signed-off-by: F. Aragon <francisco.aragon@advantech.de>

diff --git a/drivers/gpu/drm/imx/lcdifv3/lcdifv3-crtc.c b/drivers/gpu/drm/imx/lcdifv3/lcdifv3-crtc.c
index 55eec4ea74e7..e03dde6a0e9b 100644
--- a/drivers/gpu/drm/imx/lcdifv3/lcdifv3-crtc.c
+++ b/drivers/gpu/drm/imx/lcdifv3/lcdifv3-crtc.c
@@ -21,6 +21,10 @@
 #include "lcdifv3-plane.h"
 #include "lcdifv3-kms.h"
 
+// backport from https://github.com/ADVANTECH-Corp/linux-imx.git imx8LBVF0170
+// patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+#define CONFIG_ARCH_ADVANTECH
+
 struct lcdifv3_crtc {
 	struct device *dev;
 
@@ -31,6 +35,12 @@ struct lcdifv3_crtc {
 	u32 pix_fmt;		/* drm fourcc */
 };
 
+#if defined(CONFIG_ARCH_ADVANTECH) // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+static int first_flip_complete = 1;
+extern void enable_lcd_vdd_en(void);
+extern void enable_ldb_signal(void);
+#endif
+
 #define to_lcdifv3_crtc(crtc) container_of(crtc, struct lcdifv3_crtc, base)
 
 static void lcdifv3_crtc_reset(struct drm_crtc *crtc)
@@ -179,6 +189,14 @@ static void lcdifv3_crtc_atomic_enable(struct drm_crtc *crtc,
 	lcdifv3_plane_atomic_update(crtc->primary, state);
 	lcdifv3_en_shadow_load(lcdifv3);
 
+#if defined(CONFIG_ARCH_ADVANTECH) // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+	if (first_flip_complete) {
+		enable_lcd_vdd_en();
+		enable_ldb_signal();
+		first_flip_complete = 0;
+	}
+#endif
+
 	/* run LCDIFv3 */
 	lcdifv3_enable_controller(lcdifv3);
 }
diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index e5ff712bb989..a6bd8e626808 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -43,6 +43,33 @@
 #include <drm/drm_panel.h>
 #include <drm/drm_of.h>
 
+// backport from https://github.com/ADVANTECH-Corp/linux-imx.git imx8LBVF0170
+// patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+// patch2: 605789a5 Hank.Hsueh May 27 2025
+// patch3: af78e96d kuihong.xin Jun 6 2025
+#define CONFIG_ARCH_ADVANTECH
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+#include <video/mipi_display.h>
+struct dsi_ctrl_hdr {
+	u8 dtype;	/* data type */
+	u8 wait;	/* ms */
+	u8 dlen;	/* payload len */
+} __packed;
+
+struct dsi_cmd_desc {
+	struct dsi_ctrl_hdr dchdr;
+	u8 *payload;
+};
+
+struct dsi_panel_cmds {
+	u8 *buf;
+	int blen;
+	struct dsi_cmd_desc *cmds;
+	int cmd_cnt;
+};
+#endif
+
 /**
  * struct panel_desc - Describes a simple panel.
  */
@@ -162,13 +189,141 @@ struct panel_simple {
 	struct drm_display_mode override_mode;
 
 	enum drm_panel_orientation orientation;
+
+	#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	struct mipi_dsi_device *dsi;
+	struct device *dev;
+	struct dsi_panel_cmds *on_cmds;
+	struct dsi_panel_cmds *off_cmds;
+	#endif
 };
 
+#if defined(CONFIG_ARCH_ADVANTECH) // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+bool blank_flag = false;
+extern void enable_ldb_bkl_vcc(void);
+extern void enable_ldb_bkl_pwm(void);
+#endif
+
 static inline struct panel_simple *to_panel_simple(struct drm_panel *panel)
 {
 	return container_of(panel, struct panel_simple, base);
 }
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+static int panel_simple_dsi_parse_dcs_cmds(struct device *dev,
+					   const u8 *data, int blen,
+					   struct dsi_panel_cmds *pcmds)
+{
+	int len;
+	char *buf, *bp;
+	struct dsi_ctrl_hdr *dchdr;
+	int i, cnt;
+
+	if (!pcmds)
+		return -EINVAL;
+
+	buf = kmemdup(data, blen, GFP_KERNEL);
+	if (!buf)
+		return -ENOMEM;
+
+	/* scan dcs commands */
+	bp = buf;
+	len = blen;
+	cnt = 0;
+	while (len > sizeof(*dchdr)) {
+		dchdr = (struct dsi_ctrl_hdr *)bp;
+
+		if (dchdr->dlen > len) {
+			dev_err(dev, "%s: error, len=%d", __func__,
+				dchdr->dlen);
+			return -EINVAL;
+		}
+
+		bp += sizeof(*dchdr);
+		len -= sizeof(*dchdr);
+		bp += dchdr->dlen;
+		len -= dchdr->dlen;
+		cnt++;
+	}
+
+	if (len != 0) {
+		dev_err(dev, "%s: dcs_cmd=%x len=%d error!",
+			__func__, buf[0], blen);
+		kfree(buf);
+		return -EINVAL;
+	}
+
+	pcmds->cmds = kcalloc(cnt, sizeof(struct dsi_cmd_desc), GFP_KERNEL);
+	if (!pcmds->cmds) {
+		kfree(buf);
+		return -ENOMEM;
+	}
+
+	pcmds->cmd_cnt = cnt;
+	pcmds->buf = buf;
+	pcmds->blen = blen;
+
+	bp = buf;
+	len = blen;
+	for (i = 0; i < cnt; i++) {
+		dchdr = (struct dsi_ctrl_hdr *)bp;
+		len -= sizeof(*dchdr);
+		bp += sizeof(*dchdr);
+		pcmds->cmds[i].dchdr = *dchdr;
+		pcmds->cmds[i].payload = bp;
+		bp += dchdr->dlen;
+		len -= dchdr->dlen;
+	}
+
+	dev_info(dev, "%s: dcs_cmd=%x len=%d, cmd_cnt=%d\n", __func__,
+		 pcmds->buf[0], pcmds->blen, pcmds->cmd_cnt);
+	return 0;
+}
+
+static int panel_simple_dsi_send_cmds(struct panel_simple *panel,
+				      struct dsi_panel_cmds *cmds)
+{
+	struct mipi_dsi_device *dsi = panel->dsi;
+	int i, err;
+
+	//dev_err(panel->base.dev, "Warren panel_simple_dsi_send_cmds into  \n");
+	if (!cmds)
+		return -EINVAL;
+
+	for (i = 0; i < cmds->cmd_cnt; i++) {
+		struct dsi_cmd_desc *cmd = &cmds->cmds[i];
+		switch (cmd->dchdr.dtype) {
+		case MIPI_DSI_GENERIC_SHORT_WRITE_0_PARAM:
+		case MIPI_DSI_GENERIC_SHORT_WRITE_1_PARAM:
+		case MIPI_DSI_GENERIC_SHORT_WRITE_2_PARAM:
+		case MIPI_DSI_GENERIC_LONG_WRITE:
+		    //dev_err(panel->base.dev, "Warren mipi_dsi_generic_write \n");
+			err = mipi_dsi_generic_write(dsi, cmd->payload,
+						     cmd->dchdr.dlen);
+			break;
+		case MIPI_DSI_DCS_SHORT_WRITE:
+		case MIPI_DSI_DCS_SHORT_WRITE_PARAM:
+		case MIPI_DSI_DCS_LONG_WRITE:
+		    //dev_err(panel->base.dev, "Warren mipi_dsi_dcs_write_buffer \n");
+			err = mipi_dsi_dcs_write_buffer(dsi, cmd->payload,
+							cmd->dchdr.dlen);
+			break;
+		default:
+			return -EINVAL;
+		}
+
+		if (err < 0)
+			dev_err(panel->base.dev, "Warren failed to write dcs cmd: %d\n",
+				err);
+
+		if (cmd->dchdr.wait)
+			msleep(cmd->dchdr.wait);
+	}
+
+	return 0;
+}
+#endif
+
 static unsigned int panel_simple_get_timings_modes(struct panel_simple *panel,
 						   struct drm_connector *connector)
 {
@@ -318,8 +473,16 @@ static int panel_simple_suspend(struct device *dev)
 
 static int panel_simple_unprepare(struct drm_panel *panel)
 {
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	struct panel_simple *p = to_panel_simple(panel);
+#endif
 	int ret;
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	if (p->enable_gpio)
+		gpiod_set_value_cansleep(p->enable_gpio, 0);
+#endif
+
 	pm_runtime_mark_last_busy(panel->dev);
 	ret = pm_runtime_put_autosuspend(panel->dev);
 	if (ret < 0)
@@ -341,7 +504,12 @@ static int panel_simple_resume(struct device *dev)
 		return err;
 	}
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	if (p->enable_gpio)
+		gpiod_set_value_cansleep(p->enable_gpio, 1);
+#else
 	gpiod_set_value_cansleep(p->enable_gpio, 1);
+#endif
 
 	if (p->desc->delay.prepare)
 		msleep(p->desc->delay.prepare);
@@ -366,9 +534,29 @@ static int panel_simple_enable(struct drm_panel *panel)
 {
 	struct panel_simple *p = to_panel_simple(panel);
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	int err;
+#endif
+
 	if (p->desc->delay.enable)
 		msleep(p->desc->delay.enable);
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	if (p->on_cmds) {
+		err = panel_simple_dsi_send_cmds(p, p->on_cmds);
+		if (err)
+			dev_err(p->dev, "failed to send on cmds\n");
+	}
+#endif
+
+#if defined(CONFIG_ARCH_ADVANTECH) // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+	if (!blank_flag) {
+		enable_ldb_bkl_vcc();
+		enable_ldb_bkl_pwm();
+		blank_flag = true;
+	}
+#endif
+
 	return 0;
 }
 
@@ -1386,6 +1574,60 @@ static const struct panel_desc auo_t215hvn01 = {
 	.connector_type = DRM_MODE_CONNECTOR_LVDS,
 };
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+static const struct display_timing auo_g070vw01_v0_timings = {
+        .pixelclock = { 29500000, 29500000, 29500000 },
+        .hactive = { 800, 800, 800 },
+        .hfront_porch = { 24, 24, 24 },
+        .hback_porch = { 96, 96, 96 },
+        .hsync_len = { 72, 72, 72 },
+        .vactive = { 480, 480, 480 },
+        .vfront_porch = { 10, 10, 10 },
+        .vback_porch = { 3, 3, 3 },
+        .vsync_len = { 7, 7, 7 },
+        .flags = DISPLAY_FLAGS_DE_HIGH,
+};
+
+static const struct panel_desc auo_g070vw01_v0 = {
+        .timings = &auo_g070vw01_v0_timings,
+        .num_timings = 1,
+        .bpc = 8,
+        .size = {
+                .width = 152,
+                .height = 91,
+        },
+        .bus_format = MEDIA_BUS_FMT_RGB888_1X7X4_SPWG,
+        .bus_flags = DRM_BUS_FLAG_DE_HIGH,
+        .connector_type = DRM_MODE_CONNECTOR_LVDS,
+};
+
+static const struct display_timing auo_g215hvn01_0_timings = {
+        .pixelclock = { 170000000, 170000000, 170000000 },
+        .hactive = { 1920, 1920, 1920 },
+        .hfront_porch = { 30, 30, 30 },
+        .hback_porch = { 90, 90, 90 },
+        .hsync_len = { 60, 60, 60 },
+        .vactive = { 1080, 1080, 1080 },
+        .vfront_porch = { 38, 38, 38 },
+        .vback_porch = { 5, 5, 5 },
+        .vsync_len = { 7, 7, 7 },
+        .flags = DISPLAY_FLAGS_DE_HIGH,
+};
+
+static const struct panel_desc auo_g215hvn01_0 = {
+        .timings = &auo_g215hvn01_0_timings,
+        .num_timings = 1,
+        .bpc = 8,
+        .size = {
+                .width = 476,
+                .height = 268,
+        },
+        .bus_format = MEDIA_BUS_FMT_RGB888_1X7X4_SPWG,
+        .bus_flags = DRM_BUS_FLAG_DE_HIGH,
+        .connector_type = DRM_MODE_CONNECTOR_LVDS,
+};
+#endif
+
 static const struct drm_display_mode avic_tm070ddh03_mode = {
 	.clock = 51200,
 	.hdisplay = 1024,
@@ -2788,9 +3030,11 @@ static const struct panel_desc innolux_g121x1_l03 = {
 		.unprepare = 200,
 		.disable = 400,
 	},
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
 	.bus_format = MEDIA_BUS_FMT_RGB666_1X7X3_SPWG,
 	.bus_flags = DRM_BUS_FLAG_DE_HIGH,
 	.connector_type = DRM_MODE_CONNECTOR_LVDS,
+#endif
 };
 
 static const struct panel_desc innolux_g121xce_l01 = {
@@ -5070,6 +5314,14 @@ static const struct of_device_id platform_of_match[] = {
 		.compatible = "auo,t215hvn01",
 		.data = &auo_t215hvn01,
 	}, {
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+		.compatible = "auo,g070vw01_v0",
+		.data = &auo_g070vw01_v0,
+	}, {
+		.compatible = "auo,g215hvn01_0",
+		.data = &auo_g215hvn01_0,
+	}, {
+#endif
 		.compatible = "avic,tm070ddh03",
 		.data = &avic_tm070ddh03,
 	}, {
@@ -5750,11 +6002,346 @@ static const struct panel_desc_dsi osd101t2045_53ts = {
 	.lanes = 4,
 };
 
+#if defined(CONFIG_DRM_PANEL_AUO_G101UAN02) // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+// ignored (spaces/tabs): patch2: 605789a5 Hank.Hsueh May 27 2025
+static const struct drm_display_mode auo_g101uan02_mode = {
+	.clock = 148500,
+	.hdisplay = 1920,
+	.hsync_start = 1920 + 60,
+	.hsync_end = 1920 + 60 + 18,
+	.htotal = 1920 + 60 + 18 + 60,
+	.vdisplay = 1200,
+	.vsync_start = 1200 + 4,
+	.vsync_end = 1200 + 4 + 4,
+	.vtotal = 1200 + 4 + 4 + 4,
+};
+
+static const struct panel_desc_dsi auo_g101uan02 = {
+	.desc = {
+		.modes = &auo_g101uan02_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+		    .width = 217,
+		    .height = 136,
+		},
+		.connector_type = DRM_MODE_CONNECTOR_DSI,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_SYNC_PULSE,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+#endif
+
+#if defined(CONFIG_ARCH_ADVANTECH) // patch2: 605789a5 Hank.Hsueh May 27 2025
+static const struct drm_display_mode auo_g070vw01v0_mode = {
+	.clock = 29500,
+	.hdisplay = 800,
+	.hsync_start = 800 + 24,
+	.hsync_end = 800 + 24 + 72,
+	.htotal = 800 + 24 + 72 + 96,
+	.vdisplay = 480,
+	.vsync_start = 480 + 10,
+	.vsync_end = 480 + 10 + 7,
+	.vtotal = 480 + 10 + 3 + 7,
+};
+
+static const struct panel_desc_dsi auo_g070vw01v0 = {
+	.desc = {
+		.modes = &auo_g070vw01v0_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 170,
+			.height = 110,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+
+static const struct drm_display_mode auo_g150xgel05_mode = {
+	.clock = 63500,
+	.hdisplay = 1024,
+	.hsync_start = 1024 + 48,
+	.hsync_end = 1024 + 48 + 104,
+	.htotal = 1024 + 48 + 104 + 152,
+	.vdisplay = 768,
+	.vsync_start = 768 + 23,
+	.vsync_end = 768 + 23 + 4,
+	.vtotal = 768 + 23 + 3 + 4,
+};
+
+static const struct panel_desc_dsi auo_g150xgel05 = {
+	.desc = {
+		.modes = &auo_g150xgel05_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 223,
+			.height = 125,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+
+static const struct drm_display_mode auo_g215hvn01_mode = {
+	.clock = 148500,
+	.hdisplay = 1920,
+	.hsync_start = 1920 + 150,
+	.hsync_end = 1920 + 150 + 10,
+	.htotal = 1920 + 150 + 10 + 20,
+	.vdisplay = 1080,
+	.vsync_start = 1080 + 20,
+	.vsync_end = 1080 + 20 + 10,
+	.vtotal = 1080 + 20 + 20 + 10,
+};
+
+static const struct panel_desc_dsi auo_g215hvn01 = {
+	.desc = {
+		.modes = &auo_g215hvn01_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 320,
+			.height = 187,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+
+static const struct drm_display_mode innolux_r190ece_mode = {
+	.clock = 108000,
+	.hdisplay = 1280,
+	.hsync_start = 1280 + 200,
+	.hsync_end = 1280 + 200 + 10,
+	.htotal = 1280 + 200 + 10 + 188,
+	.vdisplay = 1024,
+	.vsync_start = 1024 + 20,
+	.vsync_end = 1024 + 20 + 20,
+	.vtotal = 1024 + 20 + 20 + 12,
+};
+
+static const struct panel_desc_dsi innolux_r190ece = {
+	.desc = {
+		.modes = &innolux_r190ece_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 376,
+			.height = 301,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+
+static const struct drm_display_mode lg_lp097x02_mode = {
+	.clock = 63500,
+	.hdisplay = 1024,
+	.hsync_start = 1024 + 48,
+	.hsync_end = 1024 + 48 + 104,
+	.htotal = 1024 + 48 + 104 + 152,
+	.vdisplay = 768,
+	.vsync_start = 768 + 23,
+	.vsync_end = 768 + 23 + 4,
+	.vtotal = 768 + 23 + 3 + 4,
+};
+
+static const struct panel_desc_dsi lg_lp097x02 = {
+	.desc = {
+		.modes = &lg_lp097x02_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 196,
+			.height = 147,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+
+#if defined(CONFIG_DRM_LONTIUM_LT9211)
+static const struct drm_display_mode lt9211_800x480_mode = {
+	.clock = 33000,
+	.hdisplay = 800,
+	.hsync_start = 824,
+	.hsync_end = 896,
+	.htotal = 1000,
+	.vdisplay = 480,
+	.vsync_start = 483,
+	.vsync_end = 493,
+	.vtotal = 550,
+};
+
+static const struct panel_desc_dsi lt9211_800x480 = {
+	.desc = {
+		.modes = &lt9211_800x480_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 170,
+			.height = 110,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+		.connector_type = DRM_MODE_CONNECTOR_DSI,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_SYNC_PULSE | MIPI_DSI_MODE_VIDEO_HSE | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+
+static const struct drm_display_mode lt9211_1024x768_mode = {
+	.clock = 66000,
+	.hdisplay = 1024,
+	.hsync_start = 1048,
+	.hsync_end = 1184,
+	.htotal = 1364,
+	.vdisplay = 768,
+	.vsync_start = 771,
+	.vsync_end = 777,
+	.vtotal = 806,
+};
+
+static const struct panel_desc_dsi lt9211_1024x768 = {
+	.desc = {
+		.modes = &lt9211_1024x768_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 170,
+			.height = 110,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+		.connector_type = DRM_MODE_CONNECTOR_DSI,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_SYNC_PULSE | MIPI_DSI_MODE_VIDEO_HSE | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+
+static const struct drm_display_mode lt9211_1280x800_mode = {
+	.clock = 74250,
+	.hdisplay = 1280,
+	.hsync_start = 1328,
+	.hsync_end = 1360,
+	.htotal = 1440,
+	.vdisplay = 800,
+	.vsync_start = 803,
+	.vsync_end = 809,
+	.vtotal = 859,
+};
+
+static const struct panel_desc_dsi lt9211_1280x800 = {
+	.desc = {
+		.modes = &lt9211_1280x800_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 170,
+			.height = 110,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+		.connector_type = DRM_MODE_CONNECTOR_DSI,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_SYNC_PULSE | MIPI_DSI_MODE_VIDEO_HSE | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+
+static const struct drm_display_mode lt9211_1920x1080_mode = {
+	.clock = 148500,
+	.hdisplay = 1920,
+	.hsync_start = 2008,
+	.hsync_end = 2052,
+	.htotal = 2200,
+	.vdisplay = 1080,
+	.vsync_start = 1084,
+	.vsync_end = 1089,
+	.vtotal = 1125,
+};
+
+static const struct panel_desc_dsi lt9211_1920x1080 = {
+	.desc = {
+		.modes = &lt9211_1920x1080_mode,
+		.num_modes = 1,
+		.bpc = 8,
+		.size = {
+			.width = 170,
+			.height = 110,
+		},
+		.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+		.connector_type = DRM_MODE_CONNECTOR_DSI,
+	},
+	.flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_SYNC_PULSE | MIPI_DSI_MODE_VIDEO_HSE | MIPI_DSI_MODE_NO_EOT_PACKET,
+	.format = MIPI_DSI_FMT_RGB888,
+	.lanes = 4,
+};
+#endif /*CONFIG_DRM_LONTIUM_LT9211*/
+#endif
+
+
 static const struct of_device_id dsi_of_match[] = {
 	{
 		.compatible = "auo,b080uan01",
 		.data = &auo_b080uan01
+	},
+#if defined(CONFIG_DRM_PANEL_AUO_G101UAN02) // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+	{
+		.compatible = "auo,g101uan02",
+		.data = &auo_g101uan02
+	},
+#endif
+#if defined(CONFIG_ARCH_ADVANTECH) // patch2: 605789a5 Hank.Hsueh May 27 2025
+	{
+		.compatible = "auo,g070vw01v0",
+		.data = &auo_g070vw01v0
+	}, {
+		.compatible = "auo,g150xgel05",
+		.data = &auo_g150xgel05
+	}, {
+		.compatible = "auo,g215hvn01",
+		.data = &auo_g215hvn01
+	}, {
+		.compatible = "innolux,r190ece",
+		.data = &innolux_r190ece
 	}, {
+		.compatible = "lg,lp097x02",
+		.data = &lg_lp097x02
+	},
+#if defined(CONFIG_DRM_LONTIUM_LT9211)
+	{
+		.compatible = "lontium,lt9211_800x480",
+		.data = &lt9211_800x480
+	},
+	{
+		.compatible = "lontium,lt9211_1024x768",
+		.data = &lt9211_1024x768
+	},
+	{
+		.compatible = "lontium,lt9211_1280x800",
+		.data = &lt9211_1280x800
+	},
+	{
+		.compatible = "lontium,lt9211_1920x1080",
+		.data = &lt9211_1920x1080
+	},
+#endif /*CONFIG_DRM_LONTIUM_LT9211*/
+#endif
+	{
 		.compatible = "boe,tv080wum-nl0",
 		.data = &boe_tv080wum_nl0
 	}, {
@@ -5783,6 +6370,46 @@ static int panel_simple_dsi_probe(struct mipi_dsi_device *dsi)
 	const struct panel_desc_dsi *desc;
 	struct panel_simple *panel;
 	int err;
+#if defined(CONFIG_ARCH_ADVANTECH) // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+	struct gpio_desc *gpio_dsi_vcc_enable_desc = NULL;
+	struct gpio_desc *gpio_bklt_vcc_enable_desc = NULL;
+	struct gpio_desc *gpio_dsi_bkl_enable_desc = NULL;
+#endif
+#if defined(CONFIG_ARCH_ADVANTECH) // patch2: 605789a5 Hank.Hsueh May 27 2025
+	struct gpio_desc *gpio_bklt_enable_desc = NULL;
+#endif
+#if defined(CONFIG_ARCH_ADVANTECH) // patch3: af78e96d kuihong.xin Jun 6 2025
+	const void *data;
+	int len;
+#endif
+
+#if defined(CONFIG_ARCH_ADVANTECH) // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024 // patch2: 605789a5 Hank.Hsueh May 27 2025
+	gpio_dsi_vcc_enable_desc = devm_gpiod_get_optional(&dsi->dev, "dsi-vcc-enable", GPIOD_OUT_LOW);
+	if (gpio_dsi_vcc_enable_desc != NULL)
+	{
+		gpiod_set_value_cansleep(gpio_dsi_vcc_enable_desc, 0);
+	}
+
+	gpio_bklt_vcc_enable_desc = devm_gpiod_get_optional(&dsi->dev, "bklt-vcc-enable", GPIOD_OUT_LOW);
+	if (gpio_bklt_vcc_enable_desc != NULL)
+	{
+		gpiod_set_value_cansleep(gpio_bklt_vcc_enable_desc, 0);
+	}
+
+	gpio_dsi_bkl_enable_desc = devm_gpiod_get_optional(&dsi->dev, "dsi-bkl-enable", GPIOD_OUT_LOW);
+	if (gpio_dsi_bkl_enable_desc != NULL)
+	{
+		gpiod_set_value_cansleep(gpio_dsi_bkl_enable_desc, 0);
+	}
+#endif
+
+#if defined(CONFIG_ARCH_ADVANTECH) // patch2: 605789a5 Hank.Hsueh May 27 2025
+	gpio_bklt_enable_desc = devm_gpiod_get_optional(&dsi->dev, "bklt-enable", GPIOD_OUT_LOW);
+	if (gpio_bklt_enable_desc != NULL)
+	{
+		gpiod_set_value_cansleep(gpio_bklt_enable_desc, 0);
+	}
+#endif
 
 	panel = panel_simple_probe(&dsi->dev);
 	if (IS_ERR(panel))
@@ -5793,6 +6420,47 @@ static int panel_simple_dsi_probe(struct mipi_dsi_device *dsi)
 	dsi->format = desc->format;
 	dsi->lanes = desc->lanes;
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	//--------------------------------------------------------------
+	//Warren
+
+	panel = dev_get_drvdata(&dsi->dev);
+	panel->dsi = dsi;
+
+	data = of_get_property(dsi->dev.of_node, "panel-init-sequence", &len);
+	if (data) {
+		panel->on_cmds = devm_kzalloc(&dsi->dev,
+					      sizeof(*panel->on_cmds),
+					      GFP_KERNEL);
+		if (!panel->on_cmds)
+			return -ENOMEM;
+
+		err = panel_simple_dsi_parse_dcs_cmds(&dsi->dev, data, len,
+						      panel->on_cmds);
+		if (err) {
+			dev_err(&dsi->dev, "failed to parse panel init sequence\n");
+			return err;
+		}
+	}
+
+	data = of_get_property(dsi->dev.of_node, "panel-exit-sequence", &len);
+	if (data) {
+		panel->off_cmds = devm_kzalloc(&dsi->dev,
+					       sizeof(*panel->off_cmds),
+					       GFP_KERNEL);
+		if (!panel->off_cmds)
+			return -ENOMEM;
+
+		err = panel_simple_dsi_parse_dcs_cmds(&dsi->dev, data, len,
+						      panel->off_cmds);
+		if (err) {
+			dev_err(&dsi->dev, "failed to parse panel exit sequence\n");
+			return err;
+		}
+	}
+	//--------------------------------------------------------------
+#endif
+
 	err = mipi_dsi_attach(dsi);
 	if (err) {
 		struct panel_simple *panel = mipi_dsi_get_drvdata(dsi);
diff --git a/drivers/video/backlight/pwm_bl.c b/drivers/video/backlight/pwm_bl.c
index 2d55bd5cbd32..529b851725cf 100644
--- a/drivers/video/backlight/pwm_bl.c
+++ b/drivers/video/backlight/pwm_bl.c
@@ -5,7 +5,17 @@
  * 2) platform_data being correctly configured
  */
 
+
+// backport from https://github.com/ADVANTECH-Corp/linux-imx.git imx8LBVF0170
+// patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+// patch3: af78e96d kuihong.xin Jun 6 2025
+#define CONFIG_OF
+#define CONFIG_ARCH_ADVANTECH
+
 #include <linux/delay.h>
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+#include <uapi/linux/synclink.h>
+#endif
 #include <linux/gpio/consumer.h>
 #include <linux/module.h>
 #include <linux/kernel.h>
@@ -17,6 +27,10 @@
 #include <linux/pwm_backlight.h>
 #include <linux/regulator/consumer.h>
 #include <linux/slab.h>
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+#include <linux/of_gpio.h>
+#include <linux/delay.h>
+#endif
 
 struct pwm_bl_data {
 	struct pwm_device	*pwm;
@@ -27,6 +41,9 @@ struct pwm_bl_data {
 	struct regulator	*power_supply;
 	struct gpio_desc	*enable_gpio;
 	unsigned int		scale;
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	unsigned int		dft_enable;
+#endif
 	unsigned int		post_pwm_on_delay;
 	unsigned int		pwm_off_delay;
 	int			(*notify)(struct device *,
@@ -88,12 +105,138 @@ static int compute_duty_cycle(struct pwm_bl_data *pb, int brightness, struct pwm
 	return duty_cycle + lth;
 }
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+struct gpio_desc *gpio_lvds_vcc_en_desc = NULL;
+struct gpio_desc *gpio_lvds_bkl_en_desc = NULL;
+struct gpio_desc *gpio_bklt_vcc_en_desc = NULL;
+int lvds_vcc_delay_value;
+int lvds_bkl_delay_value;
+int bklt_pwm_delay_value;
+int bklt_en_delay_value;
+int lvds_vcc_flag;
+int lvds_bkl_flag;
+int bklt_vcc_flag;
+#endif
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+int lvds_stby_flag;
+int lvds_reset_flag;
+
+struct gpio_desc *gpio_lvds_reset_en_desc = NULL;
+struct gpio_desc *gpio_lvds_stby_en_desc = NULL;
+#endif
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+void enable_lcd_vdd_en(void)
+{
+	/* LVDS Panel power enable */
+	if (gpio_lvds_vcc_en_desc != NULL)
+	{
+		printk(KERN_INFO "[LVDS Sequence] 1 Start to enable LVDS VDD. lvds_vcc_flag=%d\n",lvds_vcc_flag);
+		gpiod_set_value_cansleep(gpio_lvds_vcc_en_desc, 1);
+	}
+
+	printk(KERN_INFO "[LVDS Sequence] 2 Start to enable LVDS signal.\n");
+}
+#endif
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+void enable_bridge_stdy_en(void)
+{
+       /* Bridge IC　standby */
+	if (gpio_lvds_stby_en_desc != NULL)
+	{
+		printk(KERN_INFO "[LVDS Sequence]Bridge IC  standby and reset enable\n");
+		gpiod_set_value_cansleep(gpio_lvds_stby_en_desc, 1);
+		msleep(10);
+		gpiod_set_value_cansleep(gpio_lvds_reset_en_desc, 1);
+	}
+}
+
+void disable_bridge_stdy_en(void)
+{
+       /* Bridge IC　standby */
+	if (gpio_lvds_stby_en_desc != NULL)
+	{
+		printk(KERN_INFO "[LVDS Sequence] Bridge IC  standby and reset disable \n");
+		gpiod_set_value_cansleep(gpio_lvds_stby_en_desc, 0);
+		msleep(10);
+		gpiod_set_value_cansleep(gpio_lvds_reset_en_desc, 0);
+	}
+
+}
+#endif
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+void enable_ldb_signal(void)
+{
+	mdelay(lvds_vcc_delay_value); // T2 for AUO 7"
+}
+
+void enable_ldb_bkl_vcc(void)
+{
+	mdelay(lvds_bkl_delay_value); // T3 for AUO 7"
+
+	// Backlight On (VCC)
+	if (gpio_bklt_vcc_en_desc != NULL)
+	{
+		printk(KERN_INFO "[LVDS Sequence] 3 Start to enable backlight VCC. bklt_vcc_flag=%d\n",bklt_vcc_flag);
+		gpiod_set_value_cansleep(gpio_bklt_vcc_en_desc, 1);
+	}
+
+	mdelay(bklt_pwm_delay_value); // T8 for AUO 7"
+	printk(KERN_INFO "[LVDS Sequence] 4 Start to enable backlight PWM.\n");
+}
+
+void enable_ldb_bkl_pwm(void)
+{
+	mdelay(bklt_en_delay_value); // T9 for AUO 7"
+
+	// Backlight Enable (Display On/Off)
+	if (gpio_lvds_bkl_en_desc != NULL)
+	{
+		printk(KERN_INFO "[LVDS Sequence] 5 Start to enable LVDS backlight.\n");
+		gpiod_set_value_cansleep(gpio_lvds_bkl_en_desc, 1);
+	}
+}
+
+void disable_lcd_vdd_en(void)
+{
+	if (gpio_lvds_vcc_en_desc != NULL)
+	{
+		printk(KERN_INFO "[LVDS Sequence] 7 Start to disable LVDS VDD. lvds_vcc_flag=%d\n",lvds_vcc_flag);
+		gpiod_set_value_cansleep(gpio_lvds_vcc_en_desc, 0);
+	}
+}
+
+void disable_ldb_bkl_pwm(void)
+{
+	if (gpio_lvds_bkl_en_desc != NULL)
+	{
+		printk(KERN_INFO "[LVDS Sequence] 6 Start to disable LVDS backlight.\n");
+		gpiod_set_value_cansleep(gpio_lvds_bkl_en_desc, 0);
+	}
+}
+
+#endif
+
 static int pwm_backlight_update_status(struct backlight_device *bl)
 {
 	struct pwm_bl_data *pb = bl_get_data(bl);
+#ifndef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
 	int brightness = backlight_get_brightness(bl);
+#else
+	int brightness = bl->props.brightness;
+#endif
 	struct pwm_state state;
 
+#ifndef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	if (bl->props.power != FB_BLANK_UNBLANK ||
+	    bl->props.fb_blank != FB_BLANK_UNBLANK ||
+	    bl->props.state & BL_CORE_FBBLANK)
+		brightness = 0;
+#endif
+
 	if (pb->notify)
 		brightness = pb->notify(pb->dev, brightness);
 
@@ -357,6 +500,94 @@ static int pwm_backlight_parse_dt(struct device *dev,
 		data->max_brightness = num_levels - 1;
 	}
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+	gpio_lvds_vcc_en_desc = devm_gpiod_get_optional(dev, "lvds-vcc-enable", GPIOD_OUT_LOW);
+	gpio_bklt_vcc_en_desc = devm_gpiod_get_optional(dev, "bklt-vcc-enable", GPIOD_OUT_LOW);
+	gpio_lvds_bkl_en_desc = devm_gpiod_get_optional(dev, "lvds-bkl-enable", GPIOD_OUT_LOW);
+#endif
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	gpio_lvds_reset_en_desc= devm_gpiod_get_optional(dev, "lvds-reset",GPIOD_OUT_LOW);
+	gpio_lvds_stby_en_desc = devm_gpiod_get_optional(dev, "lvds-stby", GPIOD_OUT_LOW);
+#endif
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+	if (gpio_lvds_vcc_en_desc != NULL)
+	{
+		lvds_vcc_flag = gpiod_is_active_low(gpio_lvds_vcc_en_desc);
+	}
+
+	if (gpio_bklt_vcc_en_desc != NULL)
+	{
+		bklt_vcc_flag = gpiod_is_active_low(gpio_bklt_vcc_en_desc);
+	}
+
+	if (gpio_lvds_bkl_en_desc != NULL)
+	{
+		lvds_bkl_flag = gpiod_is_active_low(gpio_lvds_bkl_en_desc);
+	}
+#endif
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	if (gpio_lvds_stby_en_desc != NULL)
+        {
+		gpiod_set_value_cansleep(gpio_lvds_stby_en_desc, 0);
+	}
+
+	if (gpio_lvds_reset_en_desc != NULL)
+	{
+		gpiod_set_value_cansleep(gpio_lvds_reset_en_desc, 0);
+	}
+#endif
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch1: e8fcbda7 Hank.Hsueh Dec 3 2024
+	if (of_find_property(node, "skip-gpios-init", NULL)) {
+		printk("[LVDS] Skip setting GPIOs to default states\n");
+		goto get_delays;
+	}
+
+	/* Set default to output */
+	if (gpio_lvds_vcc_en_desc != NULL)
+	{
+		gpiod_set_value_cansleep(gpio_lvds_vcc_en_desc, 0);
+	}
+
+	if (gpio_bklt_vcc_en_desc != NULL)
+	{
+		gpiod_set_value_cansleep(gpio_bklt_vcc_en_desc, 0);
+	}
+
+	if (gpio_lvds_bkl_en_desc != NULL)
+        {
+		gpiod_set_value_cansleep(gpio_lvds_bkl_en_desc, 0);
+	}
+
+get_delays:
+	ret = of_property_read_u32(node,"lvds-vcc-delay-time",&lvds_vcc_delay_value);
+	if (ret < 0)
+	{
+		lvds_vcc_delay_value = 25;
+	}
+
+	ret = of_property_read_u32(node,"lvds-bkl-delay-time",&lvds_bkl_delay_value);
+	if (ret < 0)
+	{
+		lvds_bkl_delay_value = 210;
+	}
+
+	ret = of_property_read_u32(node,"bklt-pwm-delay-time",&bklt_pwm_delay_value);
+	if (ret < 0)
+	{
+		bklt_pwm_delay_value = 20;
+	}
+
+	ret = of_property_read_u32(node,"bklt-en-delay-time",&bklt_en_delay_value);
+	if (ret < 0)
+	{
+		bklt_en_delay_value = 20;
+	}
+#endif
+
 	return 0;
 }
 
@@ -460,6 +691,9 @@ static int pwm_backlight_probe(struct platform_device *pdev)
 	struct pwm_state state;
 	unsigned int i;
 	int ret;
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	struct device_node *node = pdev->dev.of_node;
+#endif
 
 	if (!data) {
 		ret = pwm_backlight_parse_dt(&pdev->dev, &defdata);
@@ -490,6 +724,13 @@ static int pwm_backlight_probe(struct platform_device *pdev)
 	pb->post_pwm_on_delay = data->post_pwm_on_delay;
 	pb->pwm_off_delay = data->pwm_off_delay;
 
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	pb->dft_enable = 1;
+	if (!of_property_read_u32(node, "default-enable", &ret)) {
+		pb->dft_enable = ret;
+	}
+#endif
+
 	pb->enable_gpio = devm_gpiod_get_optional(&pdev->dev, "enable",
 						  GPIOD_ASIS);
 	if (IS_ERR(pb->enable_gpio)) {
@@ -614,7 +855,13 @@ static int pwm_backlight_probe(struct platform_device *pdev)
 
 	bl->props.brightness = data->dft_brightness;
 	bl->props.power = pwm_backlight_initial_power_state(pb);
+
+#ifdef CONFIG_ARCH_ADVANTECH // patch3: af78e96d kuihong.xin Jun 6 2025
+	if(pb->dft_enable)
+		backlight_update_status(bl);
+#else
 	backlight_update_status(bl);
+#endif
 
 	platform_set_drvdata(pdev, bl);
 	return 0;
